### PR TITLE
Make updating label event based (superseded by #68)

### PR DESCRIPTION
**Pull request by abisammy**
This is event based way of updating the label that can be used.

However it's not fully event based, for example it has to still use the old system refreshing when no players are active, however when there is one active it is fully event based (try logging in the update method)
Have tested:
- Starting with a player active
- Opening another player 

Still need to test a few things such as:
- ~~Closing a player~~
- ~~Closing a player then opening it~~
Tested and it works 🥳

Let me know if you can think of a method to not have to search for players

### CODE REVIEW

**Comment by abisammy** [comment]
- Doesn't remove label when media player closed, need to fix later

**Comment by Moon-0xff** [comment]
Great work with this, however I definitely would prefer an implementation based around the methods and object Gjs provides.

I haven't been able to get a result using gjs signals and the other stuff but maybe you can?

Here's some links for research:  
* gjs.guide on signals: https://gjs.guide/guides/gobject/subclassing.html#signals
* gnome-shell source code (signals are implemented everywhere): https://gitlab.gnome.org/GNOME/gnome-shell/-/tree/main/
* My previous attempt to make the extension refreshless (follow the thread): https://github.com/Moon-0xff/gnome-mpris-label/issues/28#issuecomment-1445162694

**Comment by Moon-0xff** [comment]
Signals might be the wrong implementation, I'm not sure, definitely read gjs.guide for good measure.

**Comment by Batwam** [comment]
May I suggest finalising and merging the open PRs before embarking into this exercise considering how much it might be changing the underlying code base?

**Comment by Moon-0xff** [comment]
It might make more sense to merge this one first actually.  
Most of the open PRs benefit greatly from this change.

**Comment by abisammy** [comment]
TODO:
[] Conver to GObject
[] Fix bug where doesn't remove label when media player closed
[] Test and identify more bugs

**Comment by abisammy** [comment]
@Moon-0xff @Batwam Made the changes I need to this banch, I think these changes need to be tested extensively...
Would it be possible for you guys to test this. 

Please note I am aware of the bug that text is not removed when paused.

**Comment by Moon-0xff** [comment]
The diff against `main` is too big.

For this reason I avoided adding extra indentation when converting `MprisLabel` to `GObject` on 28b96aed4fb8db2119f38cca2e10ceba3691d395.

There's also way too much lines with changed indentation/style but no changed code. This further breaks `git diff` and `git blame`

Sorry, but this is just too difficult to review, can you remove the extra indentation, and revert the style changes?

**Comment by Batwam** [comment]
As a side note, I did a bit of research on the topic and found that it is possible to ignore a commit in the blame by creating a file called `.git-blame-ignore-revs` and including the hash of the commit to be ignored. 

Not sure if this is a github specific setting as I haven't tested it but if reformatting is required, this might be worth keeping im mind.

**Comment by Moon-0xff** [comment]
> As a side note, I did a bit of research on the topic and found that it is possible to ignore a commit in the blame by creating a file called .git-blame-ignore-revs and including the hash of the commit to be ignored.

Very useful. I don't think it applies here though!

If it works as I think it works then we can probably do #64 without issues.

**Comment by Batwam** [comment]
good point, I hadn't noticed #64 but that would certainly work there. My understanding is that it should be as simple as adding the file in the root directory but it would be worth checking that it works as intended. the diff will probably still show the changes which is why it would be easier to implement if there is a PR purely dedicated to it rather than combining that with other code change.

**Comment by abisammy** [comment]
Been testing this, couple bugs I noticed:

- [ ] Icon doesnt update when changing size
- [ ] Label doesnt update when changing text
- [ ] Label doesnt change when changing max size
- [ ] Label doesnt change when updating Blacklist & whitelist

Will fix in coming days

**Comment by abisammy** [comment]
Also I just noticed the update method is trigerred when switching applications, perhaps this can be disabled to further optimise performance.

**Comment by abisammy** [comment]
> Please do not add changes that doesn't pertain to the pull request, ideally the changes should be explained only by the PR title

All changes in the PR are related to removing the refresh rate, the previous commits were bug fixes for it?
Ill create a new branch from main anyway...

**Comment by abisammy** [comment]
> Sorry, it looks we aren't on the same page, my problems with the branch are explained in [this comment](https://github.com/Moon-0xff/gnome-mpris-label/pull/60#issuecomment-1589045737).
> 
> I use `git diff` and `git blame` to review and understand the submitted code, this branch doesn't play well with those commands and I can't confidently know where are the changes in the code. Therefore I can't review or merge this.

If you reopen the branch, I can make a commit to remove the indents and styling... This saves me from manually having to copy the changes...

**Comment by Moon-0xff** [comment]
Can you undo the styling changes cleanly? I'm skeptic about it.

**Comment by Moon-0xff** [comment]
Amazing!  
I will review these changes when I'm able. I was expecting a bigger diff honestly.

File: players.js [hunk 1]
```diff
@@ -41,8 +41,14 @@ const dBusInterface = `
 	</interface>
 </node>`
 
-var Players = class Players {
+var Players = 
+GObject.registerClass(
+	{
+		Signals: {'list-changed': {},}
+	},
+class Players {
```
Comment (Moon-0xff) [inline players.js:49]: shouldn't extend from `GObject.Object`?

Comment (abisammy) [reply players.js:49]: True, will fix now

**Comment by abisammy** [comment]
Merged with new changes

**Comment by abisammy** [comment]
New bug:
- [ ] For some reason when opening and closing players, the amount of events infinitley grows when firing events

**Comment by Moon-0xff** [comment]
> Mpris Label: Error: Expected type string for argument 'text' but got type boolean

Odd, perhaps the result value of `removeTextWhenPaused`?

**Comment by abisammy** [comment]
> > Mpris Label: Error: Expected type string for argument 'text' but got type boolean
> 
> Odd, perhaps the result value of `removeTextWhenPaused`?

I had thought so aswell...

I will look into it over the next few days

**Comment by abisammy** [comment]
For some reason the bug isn't being reproduced at the moment...
It could be that I had the wrong branch installed that had the bug.
I'll continue monitoring over the next few days, otherwise the feature might be working as intentionally.

**Comment by Moon-0xff** [comment]
I've been examining the changes of this branch and I've been thinking of rewriting the `Players` class so we don't need to handle the inner workings of `Players` from `extension.js` (which is a workaround, it was not intended)

I forked this branch to do this, I can't show anything useful yet.  
But the general idea is to use `proxy-changed` and a signal when `players.selected` changes to update the label/icon accordingly instead of `list-changed`

**Comment by Moon-0xff** [comment]
Branch is here under the name `refreshless`, I removed a few things, added support for GNOME <42 and made the *structure* changes I wanted.

I started *backwards* to make the structure changes first, and then connected everything. Sadly I wasn't able to make it work, I got this error:
```
JS ERROR: Error: Could not guess unspecified GValue type
update@[...]/players.js:251:8
```
And I don't know how to solve it.

**Comment by Moon-0xff** [comment]
For a clearer log, I've submitted this changes, plus the ones I've added separately as another PR (#68). 
You can, of course, add more changes by creating a PR targeting the `refreshless` branch.

Thanks @abisammy for this PR!

**Pull request closed without merging**